Sources/MongoSwift/APM.swift

```diff
@@ -0,0 +1,228 @@
+    }
+    let eventStruct = CommandSucceededEvent(event)
+    let notification = Notification(name: .commandSucceeded, userInfo: ["event": eventStruct])
+    guard let context = mongoc_apm_command_succeeded_get_context(event) else {
+        preconditionFailure("Missing context for CommandSucceededEvent")
+    }
+    let center = Unmanaged<NotificationCenter>.fromOpaque(context).takeUnretainedValue()
+    center.post(notification)
+}
+
+/// An internal callback that will be set for "command failed" events if the user
+/// enables those notifications. This function generates a new `Notification` and posts
+/// it to the NotificationCenter specified when calling `MongoClient.enableCommandMonitoring`.
+internal func commandFailed(_event: OpaquePointer?) {
+    guard let event = _event else {
+        preconditionFailure("Missing event pointer for CommandFailedEvent")
+    }
+    let eventStruct = CommandFailedEvent(event)
+    let notification = Notification(name: .commandFailed, userInfo: ["event": eventStruct])
+    guard let context = mongoc_apm_command_failed_get_context(event) else {
+        preconditionFailure("Missing context for CommandFailedEvent")
+    }
+    let center = Unmanaged<NotificationCenter>.fromOpaque(context).takeUnretainedValue()
+    center.post(notification)
+}
+
+/// Extend Notification.Name to have class properties corresponding to each type
+/// of event. This allows creating notifications and observers using the same ".x" names
+/// as those passed into `enableNotifications`.
+extension Notification.Name {
+    static let commandStarted = Notification.Name(MongoEvent.commandStarted.rawValue)
+    static let commandSucceeded = Notification.Name(MongoEvent.commandSucceeded.rawValue)
+    static let commandFailed = Notification.Name(MongoEvent.commandFailed.rawValue)
+}
+
+/// An enumeration of the events that notifications can be enabled for.
+public enum MongoEvent: String {
+    case commandStarted
+    case commandSucceeded
+    case commandFailed
+}
+
+/// An extension of MongoClient to add command monitoring and
+/// server discovery and monitoring capabilities.
+extension MongoClient {
+    /*
+     *  Enables command monitoring for this client, meaning notifications
+     *  about command events will be posted to the supplied NotificationCenter -
+     *  or if one is not provided, the default NotificationCenter.
+     *  If no specific event types are provided, all events will be posted.
+     *  Notifications can only be enabled for a single NotificationCenter at a time.
+     *
+     *  Calling this function will reset all previously enabled events - i.e.
+     *  calling
+     *      client.enableNotifications(forEvents: [.commandStarted])
+     *      client.enableNotifications(forEvents: [.commandSucceeded])
+     *
+     *  will result in only posting notifications for .commandSucceeded events.
+     */
+    public func enableCommandMonitoring(
```

Comment:
I'm wondering if we should rename this to also encompass SDAM monitoring - because the callbacks are set all at once for command monitoring + SDAM monitoring, we'll need the logic to create the `mongoc_apm_callbacks_t` in a single place. `enableApplicationPerformanceMonitoring`? or maybe just `enableMonitoring`?

Reply:
oops missed this comment before I merged, I'll rename it in the soon-to-come SDAM sketch PR